apps/telegram_bot/main.py

```diff
@@ -0,0 +1,348 @@
+for key, value in voice_attr_of.items():
+    assert value in inspect.signature(key.__init__).parameters
+
+
+# Special defaultdict that supports replitdb
+class defaultdict:
+    def __init__(self, db) -> None:
+        # Initialize an empty dictionary to store user data
+        self.db = db
+
+    def __getitem__(self, chat_id) -> Dict:
+        # Return the user data for a given user id, or create a new one if not found
+        if chat_id not in self.db:
+            self.db[chat_id] = {
+                "voices": DEFAULT_VOICES,
+                "current_voice": DEFAULT_VOICES[0],
+                "current_conversation": None,
+            }
+        return self.db[chat_id]
+
+    def __setitem__(self, chat_id, user_data) -> None:
+        # Set the user data for a given user id
+        self.db[chat_id] = user_data
+
+
+class VocodeBotResponder:
+    def __init__(
+        self,
+        transcriber: BaseTranscriber,
+        system_prompt: str,
+        synthesizer: BaseSynthesizer,
+        db=None,
+    ) -> None:
+        self.transcriber = transcriber
+        self.system_prompt = system_prompt
+        self.synthesizer = synthesizer
+        self.db = defaultdict(db if db else {})
+
+    def get_agent(self, chat_id: int) -> ChatGPTAgent:
+        # Get current voice name and description from DB
+        _, voice_name, voice_description = self.db[chat_id].get(
+            "current_voice", (None, None, None)
+        )
+
+        # Augment prompt based on available info
+        prompt = self.system_prompt
+        if voice_description != None or voice_name != None:
+            prompt += " Pretend to be {0}. This is a demo of Coqui TTS's voice creation tool, so your responses are fun, always in character, and relevant to that voice description.".format(
+                voice_description if voice_description else voice_name
+            )
+
+        # Load saved conversation if it exists
+        convo_string = self.db[chat_id]["current_conversation"]
+        agent = ChatGPTAgent(
+            system_prompt=prompt,
+            memory=pickle.loads(convo_string) if convo_string else None,
+        )
+
+        return agent
+
+    # input can be audio segment or text
+    async def get_response(
+        self, chat_id: int, input: Union[str, AudioSegment]
+    ) -> Tuple[str, AudioSegment]:
+        # If input is audio, transcribe it
+        if isinstance(input, AudioSegment):
+            input = self.transcriber.transcribe(input)
+
+        # Get agent response
+        agent = self.get_agent(chat_id)
+        agent_response = agent.respond(input)
+
+        # Set current synthesizer voice from db
+        voice_id, _, voice_description = self.db[chat_id].get(
+            "current_voice", (None, None, None)
+        )
+
+        # If we have a Coqui voice prompt, use that. Otherwise, set ID as synthesizer expects.
+        if voice_description is not None and type(self.synthesizer) == CoquiSynthesizer:
```

Comment:
nit `isinstance` instead of `type`

Reply:
done, though I'm curious: why is one better than the other?